File: src/lib/settlement.test.ts
import { calculateBalances, simplifyDebts, calculateSettlement } from './settlement';
import type { Bill } from '../schemas/session.schema';
import type { Person } from '../schemas/bill.schema';

const person = (id: string, name: string): Person => ({ id, name });

// Builds a bill with a single item split equally among the given person ids.
const makeBill = (
  id: string,
  price: number,
  consumerIds: string[],
  paidByPersonId: string | null,
  overrides: Partial<Bill> = {}
): Bill => ({
  id,
  title: `Bill ${id}`,
  date: '2026-01-01T00:00:00.000Z',
  currency: 'INR',
  taxAmount: 0,
  paidByPersonId,
  receiptImage: null,
  splitStateVersion: '2.0.0',
  items: [
    {
      id: `${id}-item`,
      name: 'Item',
      price,
      quantity: 1,
      discount: 0,
      discountType: 'flat',
      splitType: 'equal',
      consumedBy: consumerIds.map((personId) => ({ personId, value: 1 })),
    },
  ],
  ...overrides,
});

const sumBalances = (balances: { amount: number }[]) => balances.reduce((sum, b) => sum + b.amount, 0);

describe('calculateBalances', () => {
  test('single bill, single payer: payer is owed by everyone else, degenerates to a per-bill summary', () => {
    const alice = person('alice', 'Alice');
    const bob = person('bob', 'Bob');
    const carol = person('carol', 'Carol');
    const people = [alice, bob, carol];

    // 90 split 3 ways = 30 each; Alice paid, so she's owed 60 (Bob's 30 + Carol's 30)
    const bills = [makeBill('b1', 90, ['alice', 'bob', 'carol'], 'alice')];

    const balances = calculateBalances(bills, people);
    const byId = Object.fromEntries(balances.map((b) => [b.personId, b.amount]));

    expect(byId.alice).toBeCloseTo(60);
    expect(byId.bob).toBeCloseTo(-30);
    expect(byId.carol).toBeCloseTo(-30);
    expect(sumBalances(balances)).toBeCloseTo(0);
  });

  test('multiple bills, same payer: balances accumulate', () => {
    const alice = person('alice', 'Alice');
    const bob = person('bob', 'Bob');
    const people = [alice, bob];

    const bills = [makeBill('b1', 20, ['alice', 'bob'], 'alice'), makeBill('b2', 40, ['alice', 'bob'], 'alice')];

    const balances = calculateBalances(bills, people);
    const byId = Object.fromEntries(balances.map((b) => [b.personId, b.amount]));

    // Alice is owed 10 from bill 1 and 20 from bill 2 = 30
    expect(byId.alice).toBeCloseTo(30);
    expect(byId.bob).toBeCloseTo(-30);
    expect(sumBalances(balances)).toBeCloseTo(0);
  });

  test('multiple bills, different payers, with overlapping people: balances net correctly', () => {
    const alice = person('alice', 'Alice');
    const bob = person('bob', 'Bob');
    const carol = person('carol', 'Carol');
    const people = [alice, bob, carol];

    const bills = [
      makeBill('b1', 90, ['alice', 'bob', 'carol'], 'alice'), // each owes 30, alice paid
      makeBill('b2', 60, ['bob', 'carol'], 'bob'), // each owes 30, bob paid (carol only involved)
    ];

    const balances = calculateBalances(bills, people);
    const byId = Object.fromEntries(balances.map((b) => [b.personId, b.amount]));

    // Alice: +60 (b1) + 0 (not in b2) = 60
    expect(byId.alice).toBeCloseTo(60);
    // Bob: -30 (b1) + 30 (b2, owed by carol) = 0
    expect(byId.bob).toBeCloseTo(0);
    // Carol: -30 (b1) -30 (b2) = -60
    expect(byId.carol).toBeCloseTo(-60);
    expect(sumBalances(balances)).toBeCloseTo(0);
  });

  test('a person absent from some bills only accrues balance from bills they are part of', () => {
    const alice = person('alice', 'Alice');
    const bob = person('bob', 'Bob');
    const dave = person('dave', 'Dave'); // not in any bill
    const people = [alice, bob, dave];

    const bills = [makeBill('b1', 50, ['alice', 'bob'], 'alice')];

    const balances = calculateBalances(bills, people);
    const byId = Object.fromEntries(balances.map((b) => [b.personId, b.amount]));

    expect(byId.dave).toBe(0);
    expect(sumBalances(balances)).toBeCloseTo(0);
  });

  test('a bill with no payer (paidByPersonId: null) contributes nothing to any balance', () => {
    const alice = person('alice', 'Alice');
    const bob = person('bob', 'Bob');
    const people = [alice, bob];

    const bills = [makeBill('b1', 100, ['alice', 'bob'], null)];

    const balances = calculateBalances(bills, people);
    balances.forEach((b) => expect(b.amount).toBe(0));
  });

  test('zero-sum invariant holds across many bills and payers', () => {
    const people = ['a', 'b', 'c', 'd'].map((id) => person(id, id));
    const bills = [
      makeBill('b1', 37.5, ['a', 'b', 'c'], 'a'),
      makeBill('b2', 123.45, ['b', 'c', 'd'], 'c'),
      makeBill('b3', 10, ['a', 'd'], 'd'),
      makeBill('b4', 99.99, ['a', 'b', 'c', 'd'], 'b'),
    ];

    const balances = calculateBalances(bills, people);
    expect(sumBalances(balances)).toBeCloseTo(0, 6);
  });

  test('rounding: fractional-cent splits still net to zero', () => {
    const alice = person('alice', 'Alice');
    const bob = person('bob', 'Bob');
    const carol = person('carol', 'Carol');
    const people = [alice, bob, carol];

    // 10 / 3 doesn't divide evenly
    const bills = [makeBill('b1', 10, ['alice', 'bob', 'carol'], 'alice')];

    const balances = calculateBalances(bills, people);
    expect(sumBalances(balances)).toBeCloseTo(0, 2);
  });
});

describe('simplifyDebts', () => {
  test('produces zero transactions when all balances are already zero', () => {
    const transactions = simplifyDebts([
      { personId: 'a', amount: 0 },
      { personId: 'b', amount: 0 },
    ]);
    expect(transactions).toEqual([]);
  });

  test('single creditor/single debtor produces one transaction', () => {
    const transactions = simplifyDebts([
      { personId: 'a', amount: 30 },
      { personId: 'b', amount: -30 },
    ]);
    expect(transactions).toEqual([{ from: 'b', to: 'a', amount: 30 }]);
  });

  test('multiple creditors/debtors settle down to zero net balance in aggregate', () => {
    const balances = [
      { personId: 'a', amount: 60 },
      { personId: 'b', amount: 0 },
      { personId: 'c', amount: -60 },
    ];
    const transactions = simplifyDebts(balances);

    // Verify the transaction set actually zeroes out every balance
    const net: Record<string, number> = { a: 0, b: 0, c: 0 };
    transactions.forEach((t) => {
      net[t.from] -= t.amount;
      net[t.to] += t.amount;
    });
    expect(net.a).toBeCloseTo(60);
    expect(net.b).toBeCloseTo(0);
    expect(net.c).toBeCloseTo(-60);
  });

  test('ignores balances within epsilon of zero', () => {
    const transactions = simplifyDebts([
      { personId: 'a', amount: 0.0000001 },
      { personId: 'b', amount: -0.0000001 },
    ]);
    expect(transactions).toEqual([]);
  });
});

describe('calculateSettlement', () => {
  test('combines balances and transactions for a multi-payer session', () => {
    const alice = person('alice', 'Alice');
    const bob = person('bob', 'Bob');
    const carol = person('carol', 'Carol');
    const people = [alice, bob, carol];

    const bills = [makeBill('b1', 90, ['alice', 'bob', 'carol'], 'alice'), makeBill('b2', 60, ['bob', 'carol'], 'bob')];

    const result = calculateSettlement(bills, people);

    expect(sumBalances(result.balances)).toBeCloseTo(0);
    // Every transaction should be a positive amount between two distinct people
    result.transactions.forEach((t) => {
      expect(t.amount).toBeGreaterThan(0);
      expect(t.from).not.toBe(t.to);
    });
  });
});

File: server/internal/settlement/settlement_test.go
package settlement

import (
	"math"
	"testing"

	"billsplitter/server/internal/models"
)

func person(id, name string) models.Person {
	return models.Person{ID: id, Name: name}
}

// makeBill mirrors settlement.test.ts's makeBill helper: a single item split
// equally among the given consumer ids.
func makeBill(id string, price float64, consumerIDs []string, paidByPersonID *string) models.Bill {
	allocations := make([]models.Allocation, len(consumerIDs))
	for i, pid := range consumerIDs {
		allocations[i] = models.Allocation{PersonID: pid, Value: 1}
	}
	return models.Bill{
		ID:             id,
		Title:          "Bill " + id,
		Date:           "2026-01-01T00:00:00.000Z",
		Currency:       "INR",
		TaxAmount:      0,
		PaidByPersonID: paidByPersonID,
		Items: []models.Item{
			{
				ID:           id + "-item",
				Name:         "Item",
				Price:        price,
				Quantity:     1,
				Discount:     0,
				DiscountType: "flat",
				SplitType:    "equal",
				ConsumedBy:   allocations,
			},
		},
	}
}

func strPtr(s string) *string { return &s }

// makePayment mirrors settlement.test.ts's makePayment helper.
func makePayment(payerID, payeeID string, amount float64, mutate func(*models.Payment)) models.Payment {
	p := models.Payment{
		ID:        payerID + "-" + payeeID,
		PayerID:   payerID,
		PayeeID:   payeeID,
		Amount:    amount,
		Currency:  "INR",
		Method:    "cash",
		CreatedAt: "2026-01-02T00:00:00.000Z",
		Verified:  true,
	}
	verifiedAt := "2026-01-02T00:00:00.000Z"
	p.VerifiedAt = &verifiedAt
	if mutate != nil {
		mutate(&p)
	}
	return p
}

func closeTo(t *testing.T, got, want, tolerance float64) {
	t.Helper()
	if math.Abs(got-want) > tolerance {
		t.Fatalf("got %v, want %v (tolerance %v)", got, want, tolerance)
	}
}

func sumBalances(balances []Balance) float64 {
	var sum float64
	for _, b := range balances {
		sum += b.Amount
	}
	return sum
}

func balancesByID(balances []Balance) map[string]float64 {
	m := make(map[string]float64, len(balances))
	for _, b := range balances {
		m[b.PersonID] = b.Amount
	}
	return m
}

func TestCalculateBalances_SinglePayerSingleBill(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob"), person("carol", "Carol")}
	bills := []models.Bill{makeBill("b1", 90, []string{"alice", "bob", "carol"}, strPtr("alice"))}

	balances := CalculateBalances(bills, people, "INR", nil)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 60, 1e-9)
	closeTo(t, byID["bob"], -30, 1e-9)
	closeTo(t, byID["carol"], -30, 1e-9)
	closeTo(t, sumBalances(balances), 0, 1e-9)
}

func TestCalculateBalances_MultipleBillsSamePayer(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob")}
	bills := []models.Bill{
		makeBill("b1", 20, []string{"alice", "bob"}, strPtr("alice")),
		makeBill("b2", 40, []string{"alice", "bob"}, strPtr("alice")),
	}

	balances := CalculateBalances(bills, people, "INR", nil)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 30, 1e-9)
	closeTo(t, byID["bob"], -30, 1e-9)
	closeTo(t, sumBalances(balances), 0, 1e-9)
}

func TestCalculateBalances_MultipleBillsDifferentPayersOverlapping(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob"), person("carol", "Carol")}
	bills := []models.Bill{
		makeBill("b1", 90, []string{"alice", "bob", "carol"}, strPtr("alice")),
		makeBill("b2", 60, []string{"bob", "carol"}, strPtr("bob")),
	}

	balances := CalculateBalances(bills, people, "INR", nil)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 60, 1e-9)
	closeTo(t, byID["bob"], 0, 1e-9)
	closeTo(t, byID["carol"], -60, 1e-9)
	closeTo(t, sumBalances(balances), 0, 1e-9)
}

func TestCalculateBalances_PersonAbsentFromAllBills(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob"), person("dave", "Dave")}
	bills := []models.Bill{makeBill("b1", 50, []string{"alice", "bob"}, strPtr("alice"))}

	balances := CalculateBalances(bills, people, "INR", nil)
	byID := balancesByID(balances)

	if byID["dave"] != 0 {
		t.Fatalf("expected dave's balance to be 0, got %v", byID["dave"])
	}
	closeTo(t, sumBalances(balances), 0, 1e-9)
}

func TestCalculateBalances_NullPayerContributesNothing(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob")}
	bills := []models.Bill{makeBill("b1", 100, []string{"alice", "bob"}, nil)}

	balances := CalculateBalances(bills, people, "INR", nil)
	for _, b := range balances {
		if b.Amount != 0 {
			t.Fatalf("expected all balances to be 0 for a null-payer bill, got %v = %v", b.PersonID, b.Amount)
		}
	}
}

func TestCalculateBalances_ZeroSumInvariant(t *testing.T) {
	people := []models.Person{person("a", "a"), person("b", "b"), person("c", "c"), person("d", "d")}
	bills := []models.Bill{
		makeBill("b1", 37.5, []string{"a", "b", "c"}, strPtr("a")),
		makeBill("b2", 123.45, []string{"b", "c", "d"}, strPtr("c")),
		makeBill("b3", 10, []string{"a", "d"}, strPtr("d")),
		makeBill("b4", 99.99, []string{"a", "b", "c", "d"}, strPtr("b")),
	}

	balances := CalculateBalances(bills, people, "INR", nil)
	closeTo(t, sumBalances(balances), 0, 1e-6)
}

func TestCalculateBalances_FractionalCentSplitsNetToZero(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob"), person("carol", "Carol")}
	bills := []models.Bill{makeBill("b1", 10, []string{"alice", "bob", "carol"}, strPtr("alice"))}

	balances := CalculateBalances(bills, people, "INR", nil)
	closeTo(t, sumBalances(balances), 0, 1e-2)
}

func TestSimplifyDebts_AllZeroProducesNoTransactions(t *testing.T) {
	transactions := SimplifyDebts([]Balance{{PersonID: "a", Amount: 0}, {PersonID: "b", Amount: 0}})
	if len(transactions) != 0 {
		t.Fatalf("expected no transactions, got %v", transactions)
	}
}

func TestSimplifyDebts_SingleCreditorSingleDebtor(t *testing.T) {
	transactions := SimplifyDebts([]Balance{{PersonID: "a", Amount: 30}, {PersonID: "b", Amount: -30}})
	if len(transactions) != 1 {
		t.Fatalf("expected 1 transaction, got %d", len(transactions))
	}
	got := transactions[0]
	if got.From != "b" || got.To != "a" || got.Amount != 30 {
		t.Fatalf("unexpected transaction: %+v", got)
	}
}

func TestSimplifyDebts_MultipleCreditorsDebtorsNetToBalances(t *testing.T) {
	balances := []Balance{
		{PersonID: "a", Amount: 60},
		{PersonID: "b", Amount: 0},
		{PersonID: "c", Amount: -60},
	}
	transactions := SimplifyDebts(balances)

	net := map[string]float64{"a": 0, "b": 0, "c": 0}
	for _, tx := range transactions {
		net[tx.From] -= tx.Amount
		net[tx.To] += tx.Amount
	}
	closeTo(t, net["a"], 60, 1e-9)
	closeTo(t, net["b"], 0, 1e-9)
	closeTo(t, net["c"], -60, 1e-9)
}

func TestSimplifyDebts_IgnoresEpsilonBalances(t *testing.T) {
	transactions := SimplifyDebts([]Balance{{PersonID: "a", Amount: 0.0000001}, {PersonID: "b", Amount: -0.0000001}})
	if len(transactions) != 0 {
		t.Fatalf("expected no transactions, got %v", transactions)
	}
}

// makeBillWithCurrency is like makeBill but lets the caller set a currency
// different from the session's and an effective exchange rate, for the
// multi-currency conversion tests below.
func makeBillWithCurrency(id string, price float64, consumerIDs []string, paidByPersonID *string, currency string, exchangeRate *float64) models.Bill {
	bill := makeBill(id, price, consumerIDs, paidByPersonID)
	bill.Currency = currency
	bill.ExchangeRate = exchangeRate
	return bill
}

func TestCalculateBalances_ConvertsMismatchedBillCurrencyUsingExchangeRate(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob"), person("carol", "Carol")}
	rate := 80.0
	bills := []models.Bill{
		// Session currency INR; matches, no conversion.
		makeBill("b1", 90, []string{"alice", "bob", "carol"}, strPtr("alice")),
		// Bill in USD at 1 USD = 80 INR; converted before being summed.
		makeBillWithCurrency("b2", 100, []string{"bob", "carol"}, strPtr("bob"), "USD", &rate),
	}

	balances := CalculateBalances(bills, people, "INR", nil)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 60, 1e-9)
	closeTo(t, byID["bob"], -30+4000, 1e-9)
	closeTo(t, byID["carol"], -30-4000, 1e-9)
	closeTo(t, sumBalances(balances), 0, 1e-9)
}

func TestCalculateBalances_MismatchedCurrencyWithNoRateFallsBackToOne(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob")}
	bills := []models.Bill{
		makeBillWithCurrency("b1", 40, []string{"alice", "bob"}, strPtr("alice"), "USD", nil),
	}

	balances := CalculateBalances(bills, people, "INR", nil)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 20, 1e-9)
	closeTo(t, byID["bob"], -20, 1e-9)
	closeTo(t, sumBalances(balances), 0, 1e-9)
}

func TestCalculateSettlement_CombinesBalancesAndTransactions(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob"), person("carol", "Carol")}
	bills := []models.Bill{
		makeBill("b1", 90, []string{"alice", "bob", "carol"}, strPtr("alice")),
		makeBill("b2", 60, []string{"bob", "carol"}, strPtr("bob")),
	}

	result := CalculateSettlement(bills, people, "INR", nil)

	closeTo(t, sumBalances(result.Balances), 0, 1e-9)
	for _, tx := range result.Transactions {
		if tx.Amount <= 0 {
			t.Fatalf("expected positive transaction amount, got %v", tx.Amount)
		}
		if tx.From == tx.To {
			t.Fatalf("expected from != to, got both %v", tx.From)
		}
	}
}

// TestCalculateBalances_PaymentsExample mirrors settlement.test.ts's own
// worked example from changes/20260830_Payments.md: owe 4500, pay 4000
// cash, settlement shows only 500 — same fixture on both sides per
// architecture/settlement.md's two-sided-mirror discipline.
func TestCalculateBalances_PaymentsExample(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob")}
	bills := []models.Bill{makeBill("b1", 9000, []string{"alice", "bob"}, strPtr("alice"))}
	payments := []models.Payment{makePayment("bob", "alice", 4000, nil)}

	balances := CalculateBalances(bills, people, "INR", payments)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 500, 1e-9)
	closeTo(t, byID["bob"], -500, 1e-9)
	closeTo(t, sumBalances(balances), 0, 1e-9)
}

func TestCalculateBalances_UnverifiedPaymentIgnored(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob")}
	bills := []models.Bill{makeBill("b1", 9000, []string{"alice", "bob"}, strPtr("alice"))}
	payments := []models.Payment{makePayment("bob", "alice", 4000, func(p *models.Payment) {
		p.Verified = false
		p.VerifiedAt = nil
	})}

	balances := CalculateBalances(bills, people, "INR", payments)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 4500, 1e-9)
	closeTo(t, byID["bob"], -4500, 1e-9)
}

func TestCalculateBalances_PaymentInDifferentCurrencyConverted(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob")}
	bills := []models.Bill{makeBill("b1", 9000, []string{"alice", "bob"}, strPtr("alice"))}
	rate := 80.0
	// 50 USD at 80 INR/USD = 4000 INR.
	payments := []models.Payment{makePayment("bob", "alice", 50, func(p *models.Payment) {
		p.Currency = "USD"
		p.ExchangeRate = &rate
	})}

	balances := CalculateBalances(bills, people, "INR", payments)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 500, 1e-9)
	closeTo(t, byID["bob"], -500, 1e-9)
}

func TestCalculateBalances_MismatchedCurrencyPaymentWithNoRateFallsBackToOne(t *testing.T) {
	people := []models.Person{person("alice", "Alice"), person("bob", "Bob")}
	bills := []models.Bill{makeBill("b1", 9000, []string{"alice", "bob"}, strPtr("alice"))}
	payments := []models.Payment{makePayment("bob", "alice", 4000, func(p *models.Payment) {
		p.Currency = "USD"
	})}

	balances := CalculateBalances(bills, people, "INR", payments)
	byID := balancesByID(balances)

	closeTo(t, byID["alice"], 500, 1e-9)
}

func TestCalculateBalances_PaymentInvolvingOutsiderIsIgnored(t *testing.T) {
	people := []models.Person{person("alice", "Alice")}
	bills := []models.Bill{makeBill("b1", 100, []string{"alice"}, nil)}
	payments := []models.Payment{makePayment("ghost", "alice", 50, nil)}

	// Must not panic (a map write to an unknown key would).
	CalculateBalances(bills, people, "INR", payments)
}
